### PR TITLE
brew.sh: fix handling of dirty git repo

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

We cache the output of `git describe`, but we don't check whether the
cached value is valid.

This fixes that.

### CODE REVIEW

**Comment by Bo98** [comment]
I suspect this undoes some of the performance gain?

**Comment by carlocab** [comment]
Tiny bit slower, but not much:
```
❯ hyperfine --parameter-list branch master,version-cache-handling --warmup 5 --setup 'git switch {branch}' 'brew config'
Benchmark 1: brew config (branch = master)
  Time (mean ± σ):     428.1 ms ±   1.7 ms    [User: 146.5 ms, System: 117.9 ms]
  Range (min … max):   425.9 ms … 431.9 ms    10 runs

Benchmark 2: brew config (branch = version-cache-handling)
  Time (mean ± σ):     435.0 ms ±   1.7 ms    [User: 146.8 ms, System: 117.1 ms]
  Range (min … max):   433.3 ms … 438.7 ms    10 runs

Summary
  brew config (branch = master) ran
    1.02 ± 0.01 times faster than brew config (branch = version-cache-handling)
```
```
❯ hyperfine --parameter-list branch master,version-cache-handling --warmup 5 --setup 'git switch {branch}' 'brew --version'
Benchmark 1: brew --version (branch = master)
  Time (mean ± σ):      84.1 ms ±   0.7 ms    [User: 21.1 ms, System: 25.0 ms]
  Range (min … max):    82.8 ms …  85.3 ms    35 runs

Benchmark 2: brew --version (branch = version-cache-handling)
  Time (mean ± σ):      92.8 ms ±   1.2 ms    [User: 23.5 ms, System: 38.4 ms]
  Range (min … max):    91.6 ms …  96.8 ms    31 runs

Summary
  brew --version (branch = master) ran
    1.10 ± 0.02 times faster than brew --version (branch = version-cache-handling)
```

Worst-case delay is 14ms (max of `brew --version` after this change minus min of `brew --version` before this change) which is just on the edge of imperceptible.

**Comment by carlocab** [comment]
Using `read` instead of `cat`:
```
❯ hyperfine --parameter-list branch master,version-cache-handling --warmup 5 --setup 'git switch {branch}' 'brew config'
Benchmark 1: brew config (branch = master)
  Time (mean ± σ):     430.7 ms ±   3.9 ms    [User: 147.3 ms, System: 119.4 ms]
  Range (min … max):   426.6 ms … 439.9 ms    10 runs

Benchmark 2: brew config (branch = version-cache-handling)
  Time (mean ± σ):     438.3 ms ±   4.0 ms    [User: 147.9 ms, System: 119.0 ms]
  Range (min … max):   434.4 ms … 445.0 ms    10 runs

Summary
  brew config (branch = master) ran
    1.02 ± 0.01 times faster than brew config (branch = version-cache-handling)
```
```
❯ hyperfine --parameter-list branch master,version-cache-handling --warmup 5 --setup 'git switch {branch}' 'brew --version'
Benchmark 1: brew --version (branch = master)
  Time (mean ± σ):      85.1 ms ±   1.2 ms    [User: 21.2 ms, System: 25.7 ms]
  Range (min … max):    82.9 ms …  88.8 ms    34 runs

Benchmark 2: brew --version (branch = version-cache-handling)
  Time (mean ± σ):      91.8 ms ±   0.3 ms    [User: 23.4 ms, System: 37.9 ms]
  Range (min … max):    91.1 ms …  92.5 ms    31 runs

Summary
  brew --version (branch = master) ran
    1.08 ± 0.02 times faster than brew --version (branch = version-cache-handling
```

**Review by Bo98 — APPROVED** [review]
OK seems worth it for correctness

**Comment by Bo98** [comment]
Note that `git diff-index` can produce incorrect results without a `git update-index --refresh`. For example if someone edits a file and then reverts those changes (simplest reproducer: `touch $(brew --repo)/Dockerfile`).

May be worth comparing speed and correctness to `git status --porcelain --untracked-files=no`

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -443,23 +443,25 @@ GIT_REVISION=$("${HOMEBREW_GIT}" -C "${HOMEBREW_REPOSITORY}" rev-parse HEAD 2>/d
...
   unset GIT_HEAD
 fi
 
 if [[ -n "${GIT_REVISION}" ]]
 then
   GIT_DESCRIBE_CACHE_FILE="${GIT_DESCRIBE_CACHE}/${GIT_REVISION}"
-  if [[ -r "${GIT_DESCRIBE_CACHE_FILE}" ]]
+  # Pass `--` to guard against `git` being confused by a file named `HEAD`.
+  if "${HOMEBREW_GIT}" -C "${HOMEBREW_REPOSITORY}" diff-index --quiet --exit-code HEAD -- 2>/dev/null &&
+     [[ -r "${GIT_DESCRIBE_CACHE_FILE}" ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:462]:
```suggestion
  if [[ -r "${GIT_DESCRIBE_CACHE_FILE}" ]] && \
     "${HOMEBREW_GIT}" -C "${HOMEBREW_REPOSITORY}" diff --quiet HEAD -- 2>/dev/null &&
```
- much less expensive to check if the file is readable so do that first
- use `git diff` unless it's slower
- `--quiet` implies `--exit-code`

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! Would like to see reordering but using `git diff` is optional if it's slower.

File: Library/Homebrew/brew.sh (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/brew.sh:462]: If you use `git diff` then remember to pass `--no-ext-diff` too.

Comment (carlocab) [reply Library/Homebrew/brew.sh]: Done! Switched the order and used `git diff` instead.

**Pull request merged**